### PR TITLE
Multiple issues: make `offspring_dist` accept a function 

### LINKED ISSUE #25: allow `offspring` argument to be a function
the `offspring` argument to `chain_sim` should support passing a function of n (like `serial` argument).

### LINKED ISSUE #167: Consider alternative implementations for the offspring distribution
I would remove this in favour of input checking whether `roffspring_name()` returns count values. Overall I think you should restrict which distributions users can sample from, or allow this to be a parameterised anonymous function as with {cfr}.

_Originally posted by @pratikunterwegs in https://github.com/epiverse-trace/epichains/pull/122#discussion_r1420157188_

### PR DESCRIPTION
**Pull request by sbfnk**
* **Please check if the PR fulfills these requirements**

- [X] I have read the CONTRIBUTING guidelines 
- [X] A new item has been added to `NEWS.md`
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This PR changes the `offspring_dist` argument of `likelihood` and `simulate_chains` to expect a function instead of a character string.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

Yes. No version has been released yet so not going through a deprecation cycle.

* **Other information**:

Doing this necessitated two further changes:
- remove the `construct_offspring_ll_name` function and instead integrate the code into `likelihood`; because of the way the function name is extracted from the `offspring_dist` argument (using `substitute`) this can't be passed on to lower level functions; as this function was only used in this one instance I think we can live with the loss
- add an `rgborel` function; this is a difference from previous behaviour, where the `ll` function name could be informed from the string passed and now it needs a function with that name; the function itself is not actually called when estimating the likelihood if the corresponding `ll` function exists so this could, in principle, be an empty dummy function; however, for documentation/clarity purposes it is probably a good idea to have this function anyway

Closes #25 
Closes #167

### CODE REVIEW

**Comment by sbfnk** [comment]
Might need updating in line with #190

**Comment by jamesmbaazam** [comment]
LGTM. Thanks for the change. I believe it resolves #50 as well.

**Review by jamesmbaazam — APPROVED** [review]

**Comment by jamesmbaazam** [comment]
> Might need updating in line with #190

Oh, I just merged that. You can rebase on `main`.

**Comment by jamesmbaazam** [comment]
Please feel free to rebase and merge.

File: NEWS.md [hunk 1]
```diff
@@ -4,6 +4,9 @@
 a dedicated class of data structures for easy manipulation and interoperability
 with other new tools in the Epiverse and existing ecosystem.
 
+* The `offspring_dist` argument has been changed to no longer accept a character
+  string, but instead now accepts a function
```
Comment (Bisaloo) [inline NEWS.md:8]: It's probably good to motivate this change a bit more and add info about what it means in practice. Feel free to rephrase but I would like to this kind of info:

```suggestion
* The `offspring_dist` argument has been changed to no longer accept a character
  string, but instead now accepts a function (#25, #167, #188).
  This expands the number of possible functions a user can use and is no longer 
  restricted to distributions with a corresponding random number generator function
  starting with `r...()`. For example:
  
  ```r
  # This was not possible in earlier versions of this package.
  my_poisson_wrapper <- function(n) {
    rpois(n, lambda = 0.9)
  }
  chain_summary_raw <- simulate_summary(
    index_cases = 2,
    offspring_dist = my_poisson_wrapper,
    statistic = "length"
  )
  ```
```

**Review by Bisaloo — APPROVED** [review]
Thanks, it looks like a good change. I have left a couple of minor comments inline.

File: R/borel.r [hunk 2]
```diff
@@ -39,10 +39,46 @@ rborel <- function(n, mu, infinite = Inf) {
...
+##' @importFrom stats rgamma rpois
+##' @param size the dispersion parameter (often called \code{k} in ecological
+##'   applications)
+##' @param prob probability of success (in the parameterisation with
+##'   \code{prob}, see also \code{\link[stats]{NegBinomial}})
+##' @param mu mean parameter
+##' @return Vector of random numbers
+##' @author Sebastian Funk
+##' @export
+rgborel <- function(n, size, prob, mu, infinite = Inf) {
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/borel.r:62]: I'll add it to the news section instead (the function may have other uses so not convinced this info needs to be in the man page).

[the remark this answers is not in the document]
Comment (sbfnk) [reply R/borel.r:62]: Ah, thanks this makes sense. Will add that too.

[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply R/borel.r:62]: We could consider exposing the comment as a "developer" note in the function's details section. Some involved users might still want to know why it exists.

**Comment by sbfnk** [comment]
Not immediately clear to me why the R CMD CHECK action fails as it seems to run the previous version of code in the examples.

**Comment by Bisaloo** [comment]
The branch needs to be rebased and the following examples updated:

https://github.com/epiverse-trace/epichains/blob/d3e8d850bf1627aed4259af1dd7edf18beb91e41/R/epichains.R#L433-L443

https://github.com/epiverse-trace/epichains/blob/d3e8d850bf1627aed4259af1dd7edf18beb91e41/R/epichains.R#L455-L464

https://github.com/epiverse-trace/epichains/blob/d3e8d850bf1627aed4259af1dd7edf18beb91e41/R/epichains.R#L487-L505

**Comment by sbfnk** [comment]
> The branch needs to be rebased and the following examples updated:

Thanks but these are no longer present in 9953900 which failed the checks. I'm confused.

**Comment by sbfnk** [comment]
Does the action not check out a force pushed version?

**Comment by Bisaloo** [comment]
> Thanks but these are no longer present in https://github.com/epiverse-trace/epichains/commit/99539007407585f55c433e13959b2c780d651ae1 which failed the checks. I'm confused.

They are present in the merge commit since they come from https://github.com/epiverse-trace/epichains/commit/712c2661ca900fba24e4607838f337e0d46e1695 which is not included in your base branch. If you rebase your branch on top of the latest `main`, you should see them.

**Comment by sbfnk** [comment]
> > Thanks but these are no longer present in [9953900](https://github.com/epiverse-trace/epichains/commit/99539007407585f55c433e13959b2c780d651ae1) which failed the checks. I'm confused.
> 
> They are present in the merge commit since they come from [712c266](https://github.com/epiverse-trace/epichains/commit/712c2661ca900fba24e4607838f337e0d46e1695) which is not included in your base branch. If you rebase your branch on top of the latest `main`, you should see them.

Thanks - I had no idea action were checking merge commits.

**Pull request merged**